README.rst
============
dj-geocoding
============

.. image:: https://travis-ci.org/bennylope/dj-geocoding.svg?branch=master
    :target: https://travis-ci.org/bennylope/dj-geocoding

Django features for simple geocoding. Current support for the Geocodio geocoding service

Read the full docs on `Read the Docs <http://dj-geocoding.readthedocs.org/en/latest/>`_.

Installing
==========

Install dj-geocoding::

    pip install dj-geocoding

Then use it in a project::

    import dj-geocoding

Add your geocoding service API credentials to your `settings.py` file::

    GEOCODIO_API_KEY="SOMEAPIKEY"

Model
=====

Use the optional model mixin to your model if you aren't using PostGIS::

    from dj_geocoding.models import GeoBase

    class MyModel(GeoBase, models.Model):
        pass

This adds the followng fields::

    latitude = models.DecimalField(decimal_places=15, max_digits=18, null=True,
            blank=True)
    longitude = models.DecimalField(decimal_places=15, max_digits=18, null=True,
            blank=True)

Adds a geocode method::

    def geocode(self, \*args, \**kwargs):
        return geocode()

You should extend this in your model, providing the field(s) from which the
address will be pulled::

    def geocode(self):
        return super(MyModel, self).geocode('address')

You can choose the separator for providing a single address, too::

    def geocode(self):
        return super(MyModel, self).geocode('street_address', 'city', 'state',
                seperator=", ")

And additional property attributes for the `point` attribute.

Bulk geocoding
==============

The `bulk_geocode` function takes a queryset and geocodes its member objects.

.. note::

    The model *must* implement a point-type field that behaves like a
    Point field.

Example::

    geocoded_qs = bulk_geocode(MyModel.objects.all())

Specifying the field name::

    geocoded_qs = bulk_geocode(MyModel.objects.all(), field='point')

Manager
-------

The manager class implements a subclassed `QuerySet` with a `geocode` method::

    MyModel.objects.all().geocode()

This returns a queryset of the objects updated (or not) that fit within the
limits of the geocoding service. It is a convenient interface tot he
`bulk_geocode` function.

Admin site
==========

The `GeocodedFilter` filter can be used to filter locations in the admin based
on whether they have been geolocated or not.

The `GeolocateMixin` class can be added to your `ModelAdmin` definition to add
the `geocode_address` admin action. This adds the "geocode address" action to
the admin actions dropdown menu and then will allow you to geocode an entire
queryset from the admin.


HISTORY.rst
.. :changelog:

History
-------

0.2.1 (2014-12-02)
++++++++++++++++++

* Removes reliance on queryset `geocode` method in admin action
* Adds docs

0.2.0 (2014-09-08)
++++++++++++++++++

* Revise `has_geolocation` method, remove property status
* Added list filter class `GeocodedFilter`

0.1.1 (2014-08-26)
++++++++++++++++++

* Bug fix for handling single None value in _set_point

0.1.0 (2014-08-26)
++++++++++++++++++

* First release on PyPI.


tests/test_utils.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_dj-geocoding
------------

Tests for `dj-geocoding` models module.
"""

import unittest
from mock import patch, MagicMock

from dj_geocoding import utils
from test_app.models import SimpleTestModel, TestModel


class TestUtils(unittest.TestCase):

    def setUp(self):
        for i in range(3):
            SimpleTestModel.objects.create()
            TestModel.objects.create()

    @patch('dj_geocoding.utils.GeocodioClient.geocode')
    def test_geocode(self, mocked_geocode):
        mocked_result = MagicMock()
        mocked_result.coords = (12, 18)
        mocked_result.get().coords = (12, 18)
        mocked_geocode.return_value = mocked_result

        qs = utils.bulk_geocode(TestModel.objects.all())
        self.assertTrue(qs[0].latitude)


tests/test_models.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_dj-geocoding
------------

Tests for `dj-geocoding` models module.
"""

import unittest
from mock import patch, MagicMock

from decimal import Decimal
from dj_geocoding import models


class TestMixins(unittest.TestCase):

    def test_point_types(self):
        """Ensure point setter only accepts numeric types"""
        geo = models.GeoFieldsModel()
        # Floats
        geo.point = (123.12, 123.12)
        # Ints
        geo.point = (123, 123)
        # Decimals
        geo.point = (Decimal(123), Decimal(123))

    def test_none_types(self):
        """Ensure that a single NoneType value is converted"""
        geo = models.GeoFieldsModel()
        geo.point = None
        self.assertEqual(geo.point, (None, None))


    @patch('dj_geocoding.models.GeocodioClient.geocode')
    def test_geocode(self, mocked_geocode):
        """Ensure that geocoding works with various address arguments"""

        mocked_result = MagicMock()
        mocked_result.coords = (37.554895702703, -77.457561054053997)
        mocked_geocode.return_value = mocked_result

        # This could probably be mocked as well
        class PointTest(models.GeocoderMixin):
            point = None
            address = "100 main st"

        # This could probably be mocked as well
        class CoolPointTest(PointTest):
            def get_display_address(self):
                return "w00t"

        # No method, but args supplied
        geotest = PointTest()
        result = geotest.geocode('address', save=False)
        self.assertAlmostEqual(
            result[0],
            Decimal('37.554895702703'),
            places=5,
            )
        self.assertAlmostEqual(
            result[1],
            Decimal('-77.457561054053997'),
            places=5,
            )

        # If present, called
        geotest = CoolPointTest()
        result = geotest.geocode(save=False)
        self.assertAlmostEqual(
            result[0],
            Decimal('37.554895702703'),
            places=5,
            )
        self.assertAlmostEqual(
            result[1],
            Decimal('-77.457561054053997'),
            places=5,
            )

        # If not present, not called
        geotest = PointTest()
        with self.assertRaises(ValueError):
            geotest.geocode(save=False)

        # If present and args sent, args ignored
        geotest = CoolPointTest()
        geotest.geocode('kjdj', save=False)
        # Test that get_display_address was called


runtests.py
import sys

try:
    from django.conf import settings

    settings.configure(
        DEBUG=True,
        USE_TZ=True,
        DATABASES={
            "default": {
                "ENGINE": "django.db.backends.sqlite3",
            }
        },
        ROOT_URLCONF="dj_geocoding.urls",
        INSTALLED_APPS=[
            "django.contrib.auth",
            "django.contrib.contenttypes",
            "django.contrib.sites",
            "dj_geocoding",
            "test_app",
        ],
        SITE_ID=1,
        NOSE_ARGS=['-s'],
        GEOCODIO_API_KEY="blahblahblah",
    )

    try:
        import django
        setup = django.setup
    except AttributeError:
        pass
    else:
        setup()

    from django_nose import NoseTestSuiteRunner
except ImportError:
    raise ImportError("To fix this error, run: pip install -r requirements-test.txt")


def run_tests(*test_args):
    if not test_args:
        test_args = ['tests']

    # Run tests
    test_runner = NoseTestSuiteRunner(verbosity=1)

    failures = test_runner.run_tests(test_args)

    if failures:
        sys.exit(failures)


if __name__ == '__main__':
    run_tests(*sys.argv[1:])


dj_geocoding/admin.py
from django.utils.translation import ugettext_lazy as _
from django.contrib.admin import SimpleListFilter
from .utils import bulk_geocode


class GeocodedFilter(SimpleListFilter):
    """
    Admin list filter for filtering locations by whether they have
    [complete] geolocation data.
    """
    title = _('geocoded')
    parameter_name = 'geocoded'

    def lookups(self, request, model_admin):
        return (
            ('1', _('Yes')),
            ('0', _('No')),
        )

    def queryset(self, request, queryset):
        """
        Returns queryset of locations based on whether the locations
        have complete geolocation data (latitude and longitude) or
        those that lack complete geolocation data (none or only one).
        """
        if self.value() == '1':
            return queryset.filter(latitude__isnull=False, longitude__isnull=False)
        if self.value() == '0':
            return queryset.exclude(latitude__isnull=False, longitude__isnull=False)


class GeolocateMixin(object):
    """
    ModelAdmin class mixin for adding a simple geocoding interface to the
    Django admin.
    """
    actions = ['geocode_address']

    def geocode_address(self, request, queryset):
        """
        Make a request from Google via the Maps API to get the lat/lng
        locations for the selected locations.
        """
        try:
            geocoded = bulk_geocode(queryset)
        except AttributeError:
            # TODO Add a helpful error message here
            raise
        self.message_user(request,
            _("Geocoded {0} locations".format(len(geocoded))))


dj_geocoding/utils.py
from django.conf import settings
from django.db import transaction
from geocodio import GeocodioClient


def bulk_geocode(qs, *args, **kwargs):
    """
    Geocodes a queryset in bulk

    The goal is to minimize the number of API calls

    :returns: the slice of the queryset provided that was geocoded.
    """
    separator = kwargs.pop('separator', ', ')

    def location_name(location):
        # IF qs.model has attr 'get_display_address, use that
        if hasattr(qs.model, 'get_display_address'):
            return location.get_display_address()
        return separator.join([getattr(location, arg) for arg in args])

    client = GeocodioClient(settings.GEOCODIO_API_KEY)
    geocoded_addresses = client.geocode([location_name(location) for location in qs])
    with transaction.commit_on_success():
        for location in qs:
            location.point = geocoded_addresses.get(location_name(location)).coords
            location.save()
    # TODO return only portion of the qs that was geocoded
    return qs


dj_geocoding/models.py
# -*- coding: utf-8 -*-

import decimal

from django.conf import settings
from django.db import models
from django.db.models.query import QuerySet
from geocodio import GeocodioClient

from .utils import bulk_geocode


class GeocodingQuerySet(QuerySet):
    """
    QuerySet that adds a simple geocoding method.
    """
    def geocode(self, *args, **kwargs):
        """
        Geocodes all elements in the current QuerySet, setting all the given
        fields to the appropriate values.
        """
        clone = self._clone()
        return bulk_geocode(clone, *args, **kwargs)
    geocode.alters_data = True


class GeocodingManager(models.Manager):
    """
    Manager class that provides a bulk geocoding queryset method.
    """
    def get_query_set(self):
        return GeocodingQuerySet(self.model, using=self._db)

    def get_queryset(self):
        return GeocodingQuerySet(self.model, using=self._db)


class GeoFieldsModel(models.Model):
    """
    Mixin for models that require geolocation but do not make use of GeoDjagno
    features.
    """
    latitude = models.DecimalField(decimal_places=15, max_digits=18, null=True,
            blank=True)
    longitude = models.DecimalField(decimal_places=15, max_digits=18, null=True,
            blank=True)

    class Meta:
        abstract = True

    def _get_point(self):
        """
        :returns: tuple of the latitude, longitude of the point
        """
        return self.latitude, self.longitude

    def _set_point(self, point_tuple):
        """
        Method to set the latitude and longitude via a single tuple or list.
        The method verifies the data types right away, rather than just wait
        for the model's `save` method to clean the data.

        :returns: tuple of the latitude, longitude of the point
        """
        if point_tuple is None:
            return None, None
        try:
            assert len(point_tuple) == 2
        except AssertionError:
            raise Exception("point_tuple must have exactly two elements")

        self.latitude = decimal.Decimal(str(point_tuple[0]))
        self.longitude = decimal.Decimal(str(point_tuple[1]))
        return self.latitude, self.longitude

    point = property(_get_point, _set_point)

    def has_geolocation(self):
        if self.latitude is not None and self.longitude is not None:
            return True
        return False
    has_geolocation.boolean = True
    has_geolocation.short_description = "Geocoded"


class GeocoderMixin(object):

    def geocode(self, *args, **kwargs):
        """
        Geocodes the object.

        Uses either a `get_display_address` method on the object or the
        arguments provided.

        :returns: the object itself
        """
        separator = kwargs.pop('separator', ', ')
        save = kwargs.pop('save', True)
        if hasattr(self, 'get_display_address'):
            address = self.get_display_address()
        elif not args:
            raise ValueError("Must provide field names for address display if no"
                    " get_display_address method is present")
        else:
            address = separator.join([getattr(self, arg) for arg in args])
        client = GeocodioClient(settings.GEOCODIO_API_KEY)
        result = client.geocode(address)
        # TODO handle None
        self.point = decimal.Decimal(str(result.coords[0])), decimal.Decimal(str(result.coords[1]))
        if save:
            self.save()
        return self.point


class GeoBase(GeoFieldsModel, GeocoderMixin):
    """
    Base model class that provides dual fields for storing latitude and
    longitude, respectively, and methods for dealing with points and geocoding.
    """

    class Meta:
        abstract = True
